### PR TITLE
Rockmine runtime added

### PR DESCRIPTION
**Pull request by NachoPal**
Related to: #647
 
The Chain Spec json under `/polkadot-parachains/res/` are not yet updated, I guess that should be done once the runtime is reviewed and we have already at least a bootnode?

In the commit there is also a small fix here: https://github.com/paritytech/cumulus/pull/684/files#diff-dd18b092a58cf72bd304a48212a4886abe30d3ed3dbe3a46bc5e19243c6f1655R575 I think that was missconfigured

### CODE REVIEW

File: polkadot-parachains/res/rockmine.json [hunk 1]
```diff
@@ -0,0 +1,39 @@
+{
+  "name": "Westmint",
+  "id": "westmint",
```
Comment (apopiak) [inline polkadot-parachains/res/rockmine.json:3]: that seems wrong :-D
also agree on only adding this once we have a spec that we're running
as we'll upgrade shell to Rockmine we'll start with a shell Chainspec

File: polkadot-parachains/rockmine/src/constants.rs [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+pub mod fee {
+	use frame_support::weights::{
+		constants::ExtrinsicBaseWeight, WeightToFeeCoefficient, WeightToFeeCoefficients,
+		WeightToFeePolynomial,
+	};
+	use node_primitives::Balance;
+	use smallvec::smallvec;
+	pub use sp_runtime::Perbill;
+
+	/// The block saturation level. Fees will be updates based on this value.
```
Comment (apopiak) [inline polkadot-parachains/rockmine/src/constants.rs:43]:
```suggestion
	/// The block saturation level. Fees will be updated based on this value.
```

File: polkadot-parachains/rockmine/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,973 @@
...
+	CancelProxy,
+	/// Assets proxy. Can execute any call from `assets`, **including asset transfers**.
+	Assets,
+	/// Owner proxy. Can execute calls related to asset ownership.
+	AssetOwner,
+	/// Asset manager. Can execute calls related to asset management.
+	AssetManager,
+	/// Collator selection proxy. Can execute calls related to collator selection mechanism.
+	Collator,
+}
```
Comment (apopiak) [inline polkadot-parachains/rockmine/src/lib.rs:315]: I wonder whether we might want to extract this into common, probably better on a follow-up PR, though

File: polkadot-parachains/rockmine/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,973 @@
...
+	type ReservedXcmpWeight = ReservedXcmpWeight;
+}
+
+impl parachain_info::Config for Runtime {}
+
+impl cumulus_pallet_aura_ext::Config for Runtime {}
+
+parameter_types! {
+	pub const RococoLocation: MultiLocation = MultiLocation::parent();
+	pub RelayNetwork: NetworkId = NetworkId::Named(b"Rococo".to_vec());
```
Comment (apopiak) [inline polkadot-parachains/rockmine/src/lib.rs:431]:
```suggestion
	pub RelayNetwork: NetworkId = NetworkId::Polkadot;
```
Rococo has network id Polkadot https://github.com/paritytech/polkadot/blob/666e67d2e753644dfc9f042195bfc9e1051e4eb5/runtime/rococo/src/lib.rs#L603

File: polkadot-parachains/rockmine/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,973 @@
...
+);
+
+pub struct XcmConfig;
+impl Config for XcmConfig {
+	type Call = Call;
+	type XcmSender = XcmRouter;
+	type AssetTransactor = AssetTransactors;
+	type OriginConverter = XcmOriginToTransactDispatchOrigin;
+	type IsReserve = NativeAsset;
+	type IsTeleporter = NativeAsset; // <- should be enough to allow teleportation of WND
```
Comment (apopiak) [inline polkadot-parachains/rockmine/src/lib.rs:538]:
```suggestion
	type IsTeleporter = NativeAsset; // <- should be enough to allow teleportation of ROC
```

File: polkadot-parachains/rockmine/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,973 @@
...
+	// ..and XCMP to communicate with the sibling chains.
+	XcmpQueue,
+);
+
+impl pallet_xcm::Config for Runtime {
+	type Event = Event;
+	type SendXcmOrigin = EnsureXcmOrigin<Origin, LocalOriginToLocation>;
+	type XcmRouter = XcmRouter;
+	type ExecuteXcmOrigin = EnsureXcmOrigin<Origin, LocalOriginToLocation>;
+	type XcmExecuteFilter = Everything;
```
Comment (apopiak) [inline polkadot-parachains/rockmine/src/lib.rs:570]:
```suggestion
	// We just allow any XCM execution, fine for a testnet.
	type XcmExecuteFilter = Everything;
```

File: polkadot-parachains/rockmine/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,973 @@
...
+		Block = Block,
+		NodeBlock = opaque::Block,
+		UncheckedExtrinsic = UncheckedExtrinsic,
+	{
+		// System support stuff;
+		System: frame_system::{Pallet, Call, Config, Storage, Event<T>},
+		ParachainSystem: cumulus_pallet_parachain_system::{
+			Pallet, Call, Config, Storage, Inherent, Event<T>, ValidateUnsigned,
+		},
+		RandomnessCollectiveFlip: pallet_randomness_collective_flip::{Pallet, Storage},
```
Comment (apopiak) [inline polkadot-parachains/rockmine/src/lib.rs:699]: let's not add this, we'll remove it later anyway: https://github.com/paritytech/cumulus/issues/681

File: polkadot-parachains/rockmine/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,973 @@
...
+		XcmpQueue: cumulus_pallet_xcmp_queue::{Pallet, Call, Storage, Event<T>},
+		PolkadotXcm: pallet_xcm::{Pallet, Call, Event<T>, Origin},
+		CumulusXcm: cumulus_pallet_xcm::{Pallet, Event<T>, Origin},
+		DmpQueue: cumulus_pallet_dmp_queue::{Pallet, Call, Storage, Event<T>},
+
+		// More things for the main stage
+		Uniques: pallet_uniques::{Pallet, Call, Storage, Event<T>},
+
+		// More Monetary stuff
+		AssetTxPayment: pallet_asset_tx_payment::{Pallet},
```
Comment (apopiak) [inline polkadot-parachains/rockmine/src/lib.rs:731]: Let's move this up to TransactionPayment

File: polkadot-parachains/rockmine/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,973 @@
...
+		Proxy: pallet_proxy::{Pallet, Call, Storage, Event<T>},
+
+		// XCM helpers.
+		XcmpQueue: cumulus_pallet_xcmp_queue::{Pallet, Call, Storage, Event<T>},
+		PolkadotXcm: pallet_xcm::{Pallet, Call, Event<T>, Origin},
+		CumulusXcm: cumulus_pallet_xcm::{Pallet, Event<T>, Origin},
+		DmpQueue: cumulus_pallet_dmp_queue::{Pallet, Call, Storage, Event<T>},
+
+		// More things for the main stage
+		Uniques: pallet_uniques::{Pallet, Call, Storage, Event<T>},
```
Comment (apopiak) [inline polkadot-parachains/rockmine/src/lib.rs:728]:
```suggestion
		// Utilities.
		Utility: pallet_utility::{Pallet, Call, Event},
		Multisig: pallet_multisig::{Pallet, Call, Storage, Event<T>},
		Proxy: pallet_proxy::{Pallet, Call, Storage, Event<T>},
		
		// The main stage
		Assets: pallet_assets::{Pallet, Call, Storage, Event<T>},
		Uniques: pallet_uniques::{Pallet, Call, Storage, Event<T>},

		// XCM helpers.
		XcmpQueue: cumulus_pallet_xcmp_queue::{Pallet, Call, Storage, Event<T>},
		PolkadotXcm: pallet_xcm::{Pallet, Call, Event<T>, Origin},
		CumulusXcm: cumulus_pallet_xcm::{Pallet, Event<T>, Origin},
		DmpQueue: cumulus_pallet_dmp_queue::{Pallet, Call, Storage, Event<T>},
```

File: polkadot-parachains/src/chain_spec.rs [hunk 10]
```diff
@@ -713,3 +723,187 @@ fn westmint_genesis(
...
+
+	RockmineChainSpec::from_genesis(
+		// Name
+		"Rockmine",
+		// ID
+		"rockmine",
+		ChainType::Live,
+		move || {
+			rockmine_genesis(
+				// initial collators.
```
Comment (apopiak) [inline polkadot-parachains/src/chain_spec.rs:826]: how were these created?

File: polkadot-parachains/src/command.rs [hunk 11]
```diff
@@ -110,6 +118,14 @@ fn load_spec(
...
 			&include_bytes!("../res/westmint.json")[..],
 		)?),
+		"rockmine-dev" => Box::new(chain_spec::rockmine_development_config(para_id)),
+		"rockmine-local" => Box::new(chain_spec::rockmine_local_config(para_id)),
+		// the chain spec as used for generating the upgrade genesis values
+		"rockmine-genesis" => Box::new(chain_spec::rockmine_config(para_id)),
+		// the shell-based chain spec as used for syncing
+		"rockmine" => Box::new(chain_spec::ChainSpec::from_json_bytes(
+			&include_bytes!("../res/rockmine.json")[..],
+		)?),
```
Comment (apopiak) [inline polkadot-parachains/src/command.rs:128]: let's remove this until we have a chainspec for a running chain on Rococo

File: polkadot-parachains/src/chain_spec.rs [hunk 12]
```diff
@@ -188,10 +188,13 @@ pub type StatemineChainSpec =
 	sc_service::GenericChainSpec<statemine_runtime::GenesisConfig, Extensions>;
 pub type WestmintChainSpec =
 	sc_service::GenericChainSpec<westmint_runtime::GenesisConfig, Extensions>;
+pub type RockmineChainSpec =
+	sc_service::GenericChainSpec<rockmine_runtime::GenesisConfig, Extensions>;
```
Comment (apopiak) [inline polkadot-parachains/src/chain_spec.rs:192]: including another runtime will increase compile times quite a bit, we will want to look into adding conditional compilation here, similar to Polkadot https://github.com/paritytech/polkadot/blob/3bc7afae4145a06fac75e772d44a46ab824d3a93/cli/src/command.rs#L96

File: polkadot-parachains/rockmine/src/constants.rs (hunk 2 above)
Comment (NachoPal) [reply polkadot-parachains/rockmine/src/constants.rs:43]: Eagle sight 🦅 👀

Fixing it also in the rest of runtimes

**Comment by apopiak** [comment]
After further discussion we decided to deploy one of the existing chains to Rococo.

**Pull request closed without merging**